test/hotspot/gtest/opto/test_rangeinference.cpp

```diff
@@ -0,0 +1,214 @@
+#include "unittest.hpp"
+
+template <class U>
+static U uniform_random();
+
+template <>
+juint uniform_random<juint>() {
+  return os::random();
+}
+
+template <>
+julong uniform_random<julong>() {
+  return (julong(os::random()) << 32) | julong(juint(os::random()));
+}
+
+static void test_canonicalize_constraints_trivial() {
+  ASSERT_FALSE(TypeInt::NON_ZERO->contains(0));
+  ASSERT_TRUE(TypeInt::NON_ZERO->contains(1));
+  ASSERT_TRUE(TypeInt::NON_ZERO->contains(-1));
+  ASSERT_TRUE(TypeInt::CC_NE->contains(-1));
+  ASSERT_TRUE(TypeInt::CC_NE->contains(1));
+  ASSERT_FALSE(TypeInt::CC_NE->contains(0));
+  ASSERT_FALSE(TypeInt::CC_NE->contains(-2));
+  ASSERT_FALSE(TypeInt::CC_NE->contains(2));
+  ASSERT_FALSE(TypeLong::NON_ZERO->contains(jlong(0)));
+  ASSERT_TRUE(TypeLong::NON_ZERO->contains(jlong(1)));
+  ASSERT_TRUE(TypeLong::NON_ZERO->contains(jlong(-1)));
+}
+
+template <class S, class U>
+static void test_canonicalize_constraints_exhaustive() {
+  {
+    TypeIntPrototype<S, U> t{{S(0), S(0)}, {U(0), U(0)}, {U(-1), U(0)}};
+    auto new_t = t.canonicalize_constraints();
+    ASSERT_TRUE(new_t._present);
+    DEBUG_ONLY(ASSERT_TRUE(t.contains(S(0))));
+    DEBUG_ONLY(ASSERT_FALSE(t.contains(S(1))));
+  }
+  {
+    TypeIntPrototype<S, U> t{{S(0), S(0)}, {U(1), U(1)}, {U(-1), U(0)}};
+    auto new_t = t.canonicalize_constraints();
+    ASSERT_FALSE(new_t._present);
+    DEBUG_ONLY(ASSERT_FALSE(t.contains(S(0))));
+    DEBUG_ONLY(ASSERT_FALSE(t.contains(S(1))));
+  }
+  {
+    TypeIntPrototype<S, U> t{{S(S::min), S(S::max)}, {U(U::min), U(U::max)}, {U(0), U(0)}};
+    auto new_t = t.canonicalize_constraints();
+    ASSERT_TRUE(new_t._present);
+    for (int v = S::min; v <= S::max; v++) {
+      DEBUG_ONLY(ASSERT_TRUE(t.contains(S(v))));
+    }
+  }
+  for (int lo = S::min; lo <= S::max; lo++) {
+    for (int hi = lo; hi <= S::max; hi++) {
+      for (int ulo = U::min; ulo <= U::max; ulo++) {
+        for (int uhi = ulo; uhi <= U::max; uhi++) {
+          for (int zeros = U::min; zeros <= U::max; zeros++) {
+            for (int ones = U::min; ones <= U::max; ones++) {
+              TypeIntPrototype<S, U> t{{S(lo), S(hi)}, {U(ulo), U(uhi)}, {U(zeros), U(ones)}};
+              auto new_t = t.canonicalize_constraints();
+              if (new_t._present) {
+                DEBUG_ONLY(new_t._data.verify_constraints());
+              }
+              for (int v = S::min; v <= S::max; v++) {
+                if (!new_t._present) {
+                  DEBUG_ONLY(ASSERT_FALSE(t.contains(S(v))));
+                } else {
+                  DEBUG_ONLY(ASSERT_EQ(t.contains(S(v)), new_t._data.contains(S(v))));
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+template <class S, class U>
+static void test_canonicalize_constraints_simple() {
+  constexpr int parameters = 10;
+  for (int i = 0; i < parameters; i++) {
+    S a = uniform_random<U>();
+    S b = uniform_random<U>();
```

Comment:
How expensive would it be to run a few more here?

Reply:
We can run it more, but this one is the simple cases, so I thought it is not needed much, raising `parameters` to 1000 does not affect the test runtime much, so I did it anyway.